### PR TITLE
Dynamic example

### PR DESCRIPTION
**Pull request by reeceshuttle**
Im just starting this pull request to start a dialogue...its not quite ready yet. I just realized I need to still extract the init and goal strings to instead embed instead of the whole string. I also will make a few more tests for get_closest() but am pretty confident in it right now. I've got some weird errors in the tests that are coming up that im not sure of as well. Could you give me your thoughts on all of this?

### CODE REVIEW

**Comment by reeceshuttle** [comment]
Also will have to make new test cases of main for the pipeline and in the llm methods to get to 100% coverage of the new parts...

File: llm4pddl/approaches/base_approach.py [hunk 1]
```diff
@@ -14,6 +14,7 @@ class BaseApproach(abc.ABC):
 
     def __init__(self) -> None:
         self._rng = np.random.default_rng(FLAGS.seed)
+        self.embeddings_mapping: List[Dict] = []
```
Comment (tomsilver) [inline llm4pddl/approaches/base_approach.py:17]: I don't think _all_ approaches need embeddings (e.g., pure planning approaches certainly don't) -- can we just define this within the LLMOpenLoopApproach?

File: llm4pddl/approaches/llm_open_loop_approach.py [hunk 2]
```diff
@@ -22,6 +22,7 @@ def __init__(self, num_completions: int, temperature: float) -> None:
         self._llm = OpenAILLM(FLAGS.llm_model_name)
         # Set after learning.
         self._prompt_prefix = ""
+        self.embeddings_mapping: List = []
```
Comment (tomsilver) [inline llm4pddl/approaches/llm_open_loop_approach.py:25]: let's make this a private attribute (`self._embeddings_mapping`). also, "mapping" makes it sound like a dict, but this is a list? also, we'll want to make this type annotation more specific

File: llm4pddl/utils.py [hunk 3]
```diff
@@ -238,6 +241,75 @@ def minify_pddl_problem(problem: str) -> str:
...
+                            dataset: Dataset) -> List[Dict]:
+    """Makes embeddings mapping for training data."""
+    assert len(embeddings) == len(dataset)
+    return [{
+        'embedding': emb,
+        'datum': datum
+    } for emb, datum in zip(embeddings, dataset)]
+
+
+def get_closest(task: Task, embeddings_mapping: List[Dict],
```
Comment (tomsilver) [inline llm4pddl/utils.py:277]: let's name this function name more specific -- `get_closest_embeddings`?

File: llm4pddl/utils.py [hunk 4]
```diff
@@ -238,6 +241,75 @@ def minify_pddl_problem(problem: str) -> str:
...
+                num_train: int) -> List:
```
Comment (tomsilver) [inline llm4pddl/utils.py:278]: let's call this `num_embeddings` or `num_closest` or `top_n` instead of `num_train`

File: llm4pddl/main.py [hunk 5]
```diff
@@ -47,9 +47,14 @@ def _run_pipeline(approach: BaseApproach, env: BaseEnv) -> None:
     """Run training and evaluation."""
     # Run training for learning-based approaches.
     if approach.is_learning_based:
-        train_tasks = env.get_train_tasks()
+        train_tasks = env.get_train_tasks()  #list of tasks
```
Comment (tomsilver) [inline llm4pddl/main.py:50]: nit: always add a space after the hashtag

File: llm4pddl/main.py [hunk 6]
```diff
@@ -47,9 +47,14 @@ def _run_pipeline(approach: BaseApproach, env: BaseEnv) -> None:
...
         dataset = create_dataset(train_tasks)
         approach.train(dataset)
+        # Embedding the training tasks.
```
Comment (tomsilver) [inline llm4pddl/main.py:53]: rather than changing main, can we have the approach do this instead? e.g., inside of `LLMOpenLoopApproach.__init__`

File: llm4pddl/utils.py [hunk 7]
```diff
@@ -238,6 +241,75 @@ def minify_pddl_problem(problem: str) -> str:
     return new_problem
 
 
+def embed_training_tasks(training_tasks: Sequence[Task]) -> List:
+    """"embeds a list of tasks.
```
Comment (tomsilver) [inline llm4pddl/utils.py:245]: nit: capitalize the first word

File: llm4pddl/utils.py [hunk 8]
```diff
@@ -238,6 +241,75 @@ def minify_pddl_problem(problem: str) -> str:
     return new_problem
 
 
+def embed_training_tasks(training_tasks: Sequence[Task]) -> List:
```
Comment (tomsilver) [inline llm4pddl/utils.py:244]: there's nothing specific here to "training" tasks right? so we could call this `embed_tasks(tasks)`?

File: llm4pddl/utils.py [hunk 9]
```diff
@@ -238,6 +241,75 @@ def minify_pddl_problem(problem: str) -> str:
...
+    """"embeds a list of tasks.
+
+    Returns a list of embeddings with indicies corresponding to its
```
Comment (tomsilver) [inline llm4pddl/utils.py:247]: typo: indices

File: llm4pddl/approaches/base_approach.py (hunk 1 above)
Comment (tomsilver) [inline llm4pddl/approaches/base_approach.py:17]: general suggestion: add an `Embedding` type in `structs.py` and use it in all of the type annotations

File: llm4pddl/utils.py [hunk 10]
```diff
@@ -238,6 +241,75 @@ def minify_pddl_problem(problem: str) -> str:
...
+
+
+def embed_task(task: Task, embedding_model: SentenceTransformer) -> List:
+    """Embeds a task using embedding_model.
+
+    Returns a numpy array?
+    """
+    with open(task.problem_file, 'r', encoding='utf-8') as f:
+        task_string = f.read()
+    task_string = minify_pddl_problem(task_string)
```
Comment (tomsilver) [inline llm4pddl/utils.py:262]: we will want to make sure that this task string matches the task strings that we construct inside of the approach. to make sure of that, let's make a new utility function that takes the task and performs the string construction, and then use that utility function inside of both the approach and here. the new utility function should cut out code from the LLMOpenLoopApproach.

File: llm4pddl/utils.py [hunk 11]
```diff
@@ -238,6 +241,75 @@ def minify_pddl_problem(problem: str) -> str:
...
+        'embedding': emb,
+        'datum': datum
+    } for emb, datum in zip(embeddings, dataset)]
+
+
+def get_closest(task: Task, embeddings_mapping: List[Dict],
+                num_train: int) -> List:
+    """Returns the num_train most similar training tasks to the task, in
+    reverse order of similarity."""
+    embeddings_map = embeddings_mapping[:]
```
Comment (tomsilver) [inline llm4pddl/utils.py:281]: why is this needed?

File: tests/test_utils.py [hunk 12]
```diff
@@ -403,3 +539,11 @@ def test_run_planning(domain_file, problem_file, impossible_problem_file):
     with pytest.raises(NotImplementedError) as e:
         utils.run_planning(task, planner="not a real planner")
     assert "Unrecognized planner" in str(e)
+
+
+if __name__ == "__main__": # pragma: no cover
```
Comment (tomsilver) [inline tests/test_utils.py:544]: instead of this, you can test this file alone by doing `pytest -s tests/test_utils.py`. and if you want to test just a single one, you can do `pytest -s tests/test_utils.py -k test_embed_training_tasks` for example

**Review by tomsilver — COMMENTED** [review]
@reeceshuttle this is looking really good! here's a first batch of comments.

>I just realized I need to still extract the init and goal strings to instead embed instead of the whole string. 

we can defer this to a future PR (but it's good to think about how to structure the code now to make that easier in the future).

> I've got some weird errors in the tests that are coming up that im not sure of as well.

what are the errors?

File: llm4pddl/utils.py (hunk 3 above)
Comment (reeceshuttle) [reply llm4pddl/utils.py:277]: since it returns datums does get_closest_datums work?

Comment (tomsilver) [reply llm4pddl/utils.py:277]: ah yup good call!

File: llm4pddl/main.py (hunk 6 above)
Comment (reeceshuttle) [reply llm4pddl/main.py:53]: good idea, except we need the `train_tasks` and `dataset` variables in order to these operations and these are created in main in the `_run_pipeline` function. How would you go about doing that in init instead?

Comment (tomsilver) [reply llm4pddl/main.py:53]: ah good point. so then can we do it in train() instead? we can reconstruct the train tasks from the dataset, or alter train() to also take in the train tasks.

then you will probably also need to change the call to train() that you're currently adding inside the approach itself. for that, I would recommend making a helper function, like _create_prompt_prefix(), and using that instead of train(). that makes more sense anyway because it's not really like we're re-"train"-ing the model when we solve each problem.

File: llm4pddl/utils.py (hunk 10 above)
Comment (reeceshuttle) [reply llm4pddl/utils.py:262]: Just so I understand this correctly: this new function would do all the construction of the prompt that we have in OpenLoopApproach right now? would that essentially mean taking the chunks of the code in `_create_prompt` that dont need the env and implementing them in utils instead of openloopapproach?

Comment (tomsilver) [reply llm4pddl/utils.py:262]: right! an alternative would be to move this `embed_task` utility inside `LLMOpenLoopApproach` so that you can use the prompt creation code without moving it into utilities. I don't know if that would be any easier though.

Comment (reeceshuttle) [reply llm4pddl/utils.py:261]: wait im trying to fix this right now...wouldn't it be much easier to make `_create_prompt` a public method instead of a private method? Then I could simply call the function and wouldnt have to deal with moving test cases and making sure the functionality is maintained?

Comment (tomsilver) [reply llm4pddl/utils.py:261]: are you worried about calling `_create_prompt` from within a test? or where would the private-turned-public method be called?

Comment (reeceshuttle) [reply llm4pddl/utils.py:261]: wait actually never mind im just going to move `_create_prompt` to utils. that will work better I think.

Comment (tomsilver) [reply llm4pddl/utils.py:261]: okay that sounds fine!

Comment (reeceshuttle) [reply llm4pddl/utils.py:261]: actually, would that be problematic if I made `_create_prompt` public?

Comment (tomsilver) [reply llm4pddl/utils.py:261]: not necessarily, but I still would want to understand where it's going to be used outside of the class

Comment (reeceshuttle) [reply llm4pddl/utils.py:261]: Im thinking about calling it inside `embed_task` in `utils` so that the prompt string embedded is the same as the string used in prompting the model.

Comment (tomsilver) [reply llm4pddl/utils.py:261]: got it, I thought that you were considering moving `embed_task` inside of the approach. I think both `embed_task` and `create_prompt` are kinda the same category of thing, so I would recommend either making them both utilities, or both approach methods.

Comment (reeceshuttle) [reply llm4pddl/utils.py:261]: sounds good. The problem was that some other functions rely on it as well, Ill just move all the functions related to the embedding and cosine sim into approach. Ill also just move the test cases I have into the correct new location.

Comment (reeceshuttle) [reply llm4pddl/utils.py:261]: how should I access `_create_prompt` if I want to use it in a test case?

Comment (tomsilver) [reply llm4pddl/utils.py:261]: You can disable protected-access for the sake of unit tests. Search the code for "protected-access" to see an example

**Comment by reeceshuttle** [comment]
The code is pretty much done... here is what needs to be done to eliminate problems:
1. There is a weird problem in autoformat getting triggered, but when I look at the details it is not obvious what is happening. 
2. We now have 100% coverage, but `test_llm_standard_approach_dynamic_small_example` fails because of a bad api key input. This makes sense, but what doesnt is that `test_llm_standard_approach_dynamic_big_example` doesnt fail, even though it does the same operation(`approach.solve`) that the failing test case does.
3. Resolve merge conflicts. I took a peek and this seems like it will be fine. I'll do this once 1 and 2 are taken care of.

Could you look at the autoformat error and possibly provide me some insight into what is occuring? Also, any thoughts on how to get around the bad api key error(mostly just figuring out why one fails and the other doesnt)? Should we just provide a key?

**Comment by tomsilver** [comment]
re: autoformatting, sorry about that -- if you pull in the change I just merged (#115) and rerun `run_autoformat.sh`, I think it should go away.

**Comment by tomsilver** [comment]
re: the API key stuff:

* let's not add the API key to the repo itself, because (1) secrets and (2) we wouldn't want to accidentally be spending money every time the unit tests are run, if they some day start charging us for codex
* instead, let's create a "mock" version of the LLM interface, and program by hand the responses that we would like the interface to return. this is how we're testing the main LLM approaches. here's the definition of the mock class: https://github.com/Learning-and-Intelligent-Systems/llm4pddl/blob/main/tests/approaches/test_llm_open_loop_approach.py#L20 and here's where we create the mock LLM and override the real one inside the approach: https://github.com/Learning-and-Intelligent-Systems/llm4pddl/blob/main/tests/approaches/test_llm_open_loop_approach.py#L72 and here's where we set the LLM response to be whatever we want: https://github.com/Learning-and-Intelligent-Systems/llm4pddl/blob/main/tests/approaches/test_llm_open_loop_approach.py#L80
* it's strange that one of the tests is currently failing due to the API key, but the other one is not. that suggests that the second test is probably not doing what we think

**Comment by reeceshuttle** [comment]
Just got everything to work! Looks like the MockLLM usage fixed my problem with the test case. Let me know if you want me to change anything else around!

File: llm4pddl/approaches/llm_open_loop_approach.py [hunk 13]
```diff
@@ -22,6 +26,7 @@ def __init__(self, num_completions: int, temperature: float) -> None:
         self._llm = OpenAILLM(FLAGS.llm_model_name)
         # Set after learning.
         self._prompt_prefix = ""
+        self._list_embeddings_mapping: List[Dict] = []
```
Comment (tomsilver) [inline llm4pddl/approaches/llm_open_loop_approach.py:29]: Can you add a comment here briefly explaining what this is?

File: llm4pddl/approaches/llm_open_loop_approach.py [hunk 14]
```diff
@@ -182,3 +200,61 @@ def _llm_response_to_plan(response: LLMResponse, task: Task) -> Plan:
             action = f"({op} {objects_str})"
             plan.append(action)
         return plan
+
```
Comment (tomsilver) [inline llm4pddl/approaches/llm_open_loop_approach.py:203]: can we make all of these new methods private, since we don't expect to use them outside of the approach itself? (other than in tests, where we can do the disable protected-access trick)

File: llm4pddl/approaches/llm_open_loop_approach.py [hunk 15]
```diff
@@ -182,3 +200,61 @@ def _llm_response_to_plan(response: LLMResponse, task: Task) -> Plan:
...
+    def embed_task(self, task: Task,
+                   embedding_model: SentenceTransformer) -> Embedding:
+        """Embeds a task using embedding_model.
+
+        Returns a numpy array.
```
Comment (tomsilver) [inline llm4pddl/approaches/llm_open_loop_approach.py:208]: you can remove this, since we can see from the type annotation what it returns

File: llm4pddl/approaches/llm_open_loop_approach.py (hunk 13 above)
Comment (tomsilver) [reply llm4pddl/approaches/llm_open_loop_approach.py:29]: also make the type annotation more specific -- a Dict of what?

File: llm4pddl/approaches/llm_open_loop_approach.py [hunk 16]
```diff
@@ -182,3 +200,61 @@ def _llm_response_to_plan(response: LLMResponse, task: Task) -> Plan:
...
+        """
+        # note: task_string includes the Q: and A: parts, not just task string
+        task_string = self._create_prompt(task)
+        # this line is a work around for now:
```
Comment (tomsilver) [inline llm4pddl/approaches/llm_open_loop_approach.py:212]: this is a little risky if we change the Q and A stuff and forget to change things here. what would go wrong exactly if we just leave the task string alone, Q and A included? the only thing that really matters is that we're consistent about how we embed the training tasks and the eval tasks, right?

File: llm4pddl/approaches/llm_open_loop_approach.py [hunk 17]
```diff
@@ -182,3 +200,61 @@ def _llm_response_to_plan(response: LLMResponse, task: Task) -> Plan:
...
+        assert len(embeddings) == len(dataset)
+        return [{
+            'embedding': emb,
+            'datum': datum
+        } for emb, datum in zip(embeddings, dataset)]
+
+    def get_closest_datums(self, task: Task,
+                           embeddings_mapping: List[Dict[str, Any]],
+                           num_closest: int) -> List[Datum]:
+        """Returns the num_train most similar training tasks to the task, in
```
Comment (tomsilver) [inline llm4pddl/approaches/llm_open_loop_approach.py:239]: num_closest* not num_train, right? I guess these are the same thing right now, but won't necessarily be in the near future

File: llm4pddl/approaches/llm_open_loop_approach.py [hunk 18]
```diff
@@ -182,3 +200,61 @@ def _llm_response_to_plan(response: LLMResponse, task: Task) -> Plan:
...
+        reverse order of similarity."""
```
Comment (tomsilver) [inline llm4pddl/approaches/llm_open_loop_approach.py:240]: a little ambiguous -- can we say "in order from least to most similar"?

File: llm4pddl/approaches/llm_open_loop_approach.py [hunk 19]
```diff
@@ -182,3 +200,61 @@ def _llm_response_to_plan(response: LLMResponse, task: Task) -> Plan:
...
+        assert num_closest <= len(embeddings_mapping)
+        embedding_model = SentenceTransformer(FLAGS.embedding_model_name)
```
Comment (tomsilver) [inline llm4pddl/approaches/llm_open_loop_approach.py:242]: can we create this model once and save it as `self._embedding_model` in `__init__()`? I don't know if it's costly to instantiate, but might as well just do it once if we can

Comment (tomsilver) [reply llm4pddl/approaches/llm_open_loop_approach.py:242]: and if we do this, you could potentially remove the `embedding_model` arguments from the helper methods and just use `self._embedding_model` within those methods

File: llm4pddl/approaches/llm_open_loop_approach.py [hunk 20]
```diff
@@ -182,3 +200,61 @@ def _llm_response_to_plan(response: LLMResponse, task: Task) -> Plan:
...
+        # now compare this embedding to all the other embeddings
+        other_embeddings = [
+            mapping['embedding'] for mapping in embeddings_mapping
+        ]
+        cos_sims = [
+            self.get_cosine_sim(task_embedding, other_emb)
+            for other_emb in other_embeddings
+        ]
+        # we will need a get_init_string() and get_goal_string() helper funcs.
+        indicies = np.argsort(cos_sims)[-num_closest:]
```
Comment (tomsilver) [inline llm4pddl/approaches/llm_open_loop_approach.py:253]: typo: indices

**Review by tomsilver — COMMENTED** [review]
really awesome work! here are some suggestions. also, for my understanding: the expected behavior right now if we were to set `--use_dynamic_examples` is that things would pretty much be the same, except the training examples are gonna be ordered in the prompt from least to most similar. is that right? and if so, we're planning to soon introduce a new flag that will allow us to make the number of training problems different from and greater than the number of prompting problems, yeah?

File: llm4pddl/approaches/llm_open_loop_approach.py (hunk 19 above)
Comment (reeceshuttle) [reply llm4pddl/approaches/llm_open_loop_approach.py:242]: good idea

File: llm4pddl/approaches/llm_open_loop_approach.py [hunk 21]
```diff
@@ -22,6 +26,11 @@ def __init__(self, num_completions: int, temperature: float) -> None:
         self._llm = OpenAILLM(FLAGS.llm_model_name)
         # Set after learning.
         self._prompt_prefix = ""
+        # _list_embeddings_mapping is a list of dictionaries of the form
```
Comment (tomsilver) [inline llm4pddl/approaches/llm_open_loop_approach.py:29]: nice!

File: llm4pddl/approaches/llm_open_loop_approach.py [hunk 22]
```diff
@@ -22,6 +26,11 @@ def __init__(self, num_completions: int, temperature: float) -> None:
...
+        # {'embedding': Embedding, 'datum': Datum}, representing a Datum
+        # in the training set and its that datum's task string's embedding.
+        self._list_embeddings_mapping: List[Dict[str, Any]] = []
+        self.embedding_model = SentenceTransformer(FLAGS.embedding_model_name)
```
Comment (tomsilver) [inline llm4pddl/approaches/llm_open_loop_approach.py:33]: let's make this private too

**Review by tomsilver — APPROVED** [review]
one or two more tiny things, otherwise lgtm!

File: llm4pddl/approaches/llm_open_loop_approach.py (hunk 22 above)
Comment (reeceshuttle) [reply llm4pddl/approaches/llm_open_loop_approach.py]: done!

**Comment by reeceshuttle** [comment]
Im seeing a coverage file in my changed files. Should that be there?

**Comment by tomsilver** [comment]
> Im seeing a coverage file in my changed files. Should that be there?

sorry I was trigger happy with merging, fixing here: https://github.com/Learning-and-Intelligent-Systems/llm4pddl/pull/129
